File: python/tests/test_abit_star.py
"""ABIT*: inflated/truncated batch planner — reaches the goal, fails when walled
off, respects param validation, converges to a near-optimal (BIT*-grade) path as
its ε schedule relaxes to 1, and keeps the anytime property under heavy inflation."""

from __future__ import annotations

from pathlib import Path

import pytest
import yaml
from conftest import CONFIG_DIR, config, grid_from, open_grid

from navigation.core.params import ParamError, ParamSet
from navigation.global_planning import RRT
from navigation.global_planning.sampling import ABITStar
from navigation.maps.occupancy_grid import OccupancyGrid2D

# Small batch budget: still reaches the goal on the open grid while staying fast.
_BATCH = {"batch_size": 150, "max_batches": 4}


def _config(tmp_path: Path, overrides: dict[str, float]) -> ParamSet:
    doc = yaml.safe_load((CONFIG_DIR / "abit_star.yaml").read_text())
    for entry in doc["params"]:
        if entry["name"] in overrides:
            entry["default"] = overrides[entry["name"]]
    out = tmp_path / "abit_star.yaml"
    out.write_text(yaml.safe_dump(doc), encoding="utf-8")
    return ParamSet.from_yaml(out)


def _motion_valid_path(grid: OccupancyGrid2D, path: list[tuple[float, float]]) -> bool:
    return all(grid.is_motion_valid(a, b) for a, b in zip(path, path[1:], strict=False))


def test_reaches_goal_on_open_grid(tmp_path: Path) -> None:
    grid = open_grid(10, 10, seed=7)
    start, goal = (0.5, 0.5), (9.0, 9.0)
    result = ABITStar(_config(tmp_path, _BATCH)).plan(grid, start, goal)
    assert result.success
    # start/goal are permanent samples, so the path pins to them exactly.
    assert result.path[0] == pytest.approx(start)
    assert result.path[-1] == pytest.approx(goal)
    assert _motion_valid_path(grid, result.path)
    assert result.cost > 0.0


def test_no_path_when_walled_off(tmp_path: Path) -> None:
    grid = grid_from(["..#..", "..#..", "..#..", "..#..", "..#.."], seed=3)
    result = ABITStar(_config(tmp_path, _BATCH)).plan(grid, (0.5, 0.5), (4.5, 4.5))
    assert not result.success
    assert result.path == []
    assert result.cost == 0.0


def test_near_optimal_after_deflation(tmp_path: Path) -> None:
    # The ε schedule relaxes to 1 on the last batch, so ABIT* recovers BIT*'s
    # admissible search and its cost sits close to the straight-line lower bound
    # even when the first batch was heavily inflated.
    grid = open_grid(12, 12, seed=5)
    start, goal = (0.5, 0.5), (11.0, 11.0)
    lower_bound = grid.distance(start, goal)
    result = ABITStar(_config(tmp_path, {**_BATCH, "inflation_factor": 100.0})).plan(
        grid, start, goal
    )
    assert result.success
    assert result.cost <= lower_bound * 1.15


def test_no_worse_than_plain_rrt(tmp_path: Path) -> None:
    # ABIT* is asymptotically optimal; a plain RRT returns the first feasible path.
    # On the same scenario+budget ABIT*'s path must be no longer than RRT's.
    grid = open_grid(12, 12, seed=11)
    start, goal = (0.5, 0.5), (11.0, 11.0)
    rrt = RRT(config("rrt")).plan(grid, start, goal)
    abit = ABITStar(_config(tmp_path, _BATCH)).plan(grid, start, goal)
    assert rrt.success and abit.success
    assert abit.cost <= rrt.cost + 1e-6


def test_anytime_under_heavy_inflation(tmp_path: Path) -> None:
    # Anytime property: with a large ε_infl held across a single batch (no
    # de-inflation), the greedy-toward-goal ordering still returns a feasible,
    # collision-free path — inflation only reorders work, never blocks a solution.
    grid = open_grid(12, 12, seed=9)
    start, goal = (0.5, 0.5), (11.0, 11.0)
    cfg = {"batch_size": 300, "max_batches": 1, "inflation_factor": 1000.0}
    result = ABITStar(_config(tmp_path, cfg)).plan(grid, start, goal)
    assert result.success
    assert result.path[-1] == pytest.approx(goal)
    assert _motion_valid_path(grid, result.path)


def test_bad_default_param_throws(tmp_path: Path) -> None:
    doc = yaml.safe_load((CONFIG_DIR / "abit_star.yaml").read_text())
    for entry in doc["params"]:
        if entry["name"] == "inflation_factor":
            entry["default"] = 0.5  # below declared min 1.0
    out = tmp_path / "abit_star.yaml"
    out.write_text(yaml.safe_dump(doc), encoding="utf-8")
    with pytest.raises(ParamError):
        ParamSet.from_yaml(out)

File: cpp/tests/test_abit_star.cpp
#include <cmath>
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "navigation/global_planning/sampling/abit_star.hpp"
#include "navigation/global_planning/sampling/rrt.hpp"
#include "test_util.hpp"

using namespace navigation;
using core::Point;

namespace {

std::vector<std::string> open_rows(int n = 10) {
  return std::vector<std::string>(static_cast<size_t>(n), std::string(static_cast<size_t>(n), '.'));
}

// Small batch budget over the open 5m x 5m grid: reliably connects start/goal
// while staying fast. `inflation` lets a test hold the schedule at a chosen value.
std::string abit_star_cfg(double inflation = 10.0, int max_batches = 5) {
  return test::write_temp(
      "abit_star.yaml",
      "algorithm: abit_star\ncategory: global_planning\nparams:\n"
      "  - {name: batch_size, type: int, default: 120, min: 1, max: 100000, description: b}\n"
      "  - {name: max_batches, type: int, default: " +
          std::to_string(max_batches) +
          ", min: 1, max: 10000, description: mb}\n"
          "  - {name: gamma, type: float, default: 30.0, min: 0.01, max: 1000, description: g}\n"
          "  - {name: inflation_factor, type: float, default: " +
          std::to_string(inflation) +
          ", min: 1.0, max: 1000000, description: ei}\n"
          "  - {name: inflation_final, type: float, default: 1.0, min: 1.0, max: 1000000, description: eif}\n"
          "  - {name: truncation_factor, type: float, default: 2.0, min: 1.0, max: 1000000, description: et}\n"
          "  - {name: seed, type: int, default: 7, min: 0, max: 2147483647, description: s}\n");
}

std::string rrt_cfg() {
  return test::write_temp("rrt.yaml",
                          "algorithm: rrt\ncategory: global_planning\nparams:\n"
                          "  - {name: max_iterations, type: int, default: 4000, min: 1, max: 200000, description: n}\n"
                          "  - {name: step_size, type: float, default: 0.5, min: 0.01, max: 100, description: eta}\n"
                          "  - {name: goal_bias, type: float, default: 0.2, min: 0.0, max: 1.0, description: b}\n"
                          "  - {name: goal_tolerance, type: float, default: 0.3, min: 0.0, max: 100, description: t}\n"
                          "  - {name: seed, type: int, default: 7, min: 0, max: 2147483647, description: s}\n");
}

double straight_line(const Point& a, const Point& b) {
  double dx = a.x - b.x, dy = a.y - b.y;
  return std::sqrt(dx * dx + dy * dy);
}

}  // namespace

// (a) valid, near-optimal path on an open grid --------------------------------

TEST(AbitStar, FindsNearOptimalPath) {
  auto g = test::make_grid(open_rows());
  Point start{0.25, 0.25}, goal{4.75, 4.75};
  global_planning::AbitStarPlanner p(core::ParamSet::from_yaml(abit_star_cfg()));
  auto r = p.plan(g, start, goal, nullptr);
  ASSERT_TRUE(r.success);
  ASSERT_GE(r.path.size(), 2u);
  EXPECT_NEAR(r.path.front().x, start.x, 1e-9);
  EXPECT_NEAR(r.path.front().y, start.y, 1e-9);
  // start/goal are permanent samples, so the path pins to the exact goal.
  EXPECT_NEAR(r.path.back().x, goal.x, 1e-9);
  EXPECT_NEAR(r.path.back().y, goal.y, 1e-9);
  for (size_t i = 0; i + 1 < r.path.size(); ++i) {
    EXPECT_TRUE(g.is_motion_valid(r.path[i], r.path[i + 1])) << "segment must be collision-free";
  }
  // The ε schedule relaxes to 1 on the last batch, so ABIT* recovers BIT*'s
  // admissible search and stays close to the straight-line lower bound.
  EXPECT_LE(r.cost, straight_line(start, goal) * 1.15);
}

// ABIT* is asymptotically optimal; a plain RRT returns the first feasible path.
// On the same scenario+budget ABIT*'s path must be no longer than RRT's.
TEST(AbitStar, NoWorseThanPlainRrt) {
  auto g = test::make_grid(open_rows());
  Point start{0.25, 0.25}, goal{4.75, 4.75};
  global_planning::AbitStarPlanner abit(core::ParamSet::from_yaml(abit_star_cfg()));
  auto ra = abit.plan(g, start, goal, nullptr);
  global_planning::RrtPlanner rrt(core::ParamSet::from_yaml(rrt_cfg()));
  auto rr = rrt.plan(g, start, goal, nullptr);
  ASSERT_TRUE(ra.success);
  ASSERT_TRUE(rr.success);
  EXPECT_LE(ra.cost, rr.cost + 1e-6);
}

// inflation > 1 anytime property: a large ε_infl held across a single batch (no
// de-inflation) still returns a feasible, collision-free path — inflation only
// reorders work, never blocks a solution.
TEST(AbitStar, AnytimeUnderHeavyInflation) {
  auto g = test::make_grid(open_rows());
  Point start{0.25, 0.25}, goal{4.75, 4.75};
  global_planning::AbitStarPlanner p(core::ParamSet::from_yaml(abit_star_cfg(1000.0, 1)));
  auto r = p.plan(g, start, goal, nullptr);
  ASSERT_TRUE(r.success);
  EXPECT_NEAR(r.path.back().x, goal.x, 1e-9);
  EXPECT_NEAR(r.path.back().y, goal.y, 1e-9);
  for (size_t i = 0; i + 1 < r.path.size(); ++i) {
    EXPECT_TRUE(g.is_motion_valid(r.path[i], r.path[i + 1])) << "segment must be collision-free";
  }
}

// (b) no-path case: a full wall separates start from goal ---------------------

TEST(AbitStar, NoPathAcrossFullWall) {
  auto rows = open_rows();
  rows[5] = "##########";  // impassable horizontal wall
  auto g = test::make_grid(rows);
  Point start{0.25, 0.25}, goal{4.75, 4.75};
  global_planning::AbitStarPlanner p(core::ParamSet::from_yaml(abit_star_cfg()));
  auto r = p.plan(g, start, goal, nullptr);
  EXPECT_FALSE(r.success);
  EXPECT_TRUE(r.path.empty());
  EXPECT_EQ(r.cost, 0.0);
}

// (c) param validation failure ------------------------------------------------

TEST(AbitStar, BadInflationThrows) {
  std::string bad = test::write_temp(
      "abit_star.yaml",
      "algorithm: abit_star\ncategory: global_planning\nparams:\n"
      "  - {name: inflation_factor, type: float, default: 0.5, min: 1.0, max: 1000000, description: under min}\n");
  EXPECT_THROW(core::ParamSet::from_yaml(bad), std::runtime_error);
}
